src/Spatial/Euclidean/StraightLine2D.cs

```diff
@@ -0,0 +1,315 @@
+            {
+                yoffset = double.NaN; // special case vertical line has no y-intercept
+                xoffset = p1.X;
+            }
+            else
+            {
+                yoffset = p1.Y - (Gradient * p1.X);
+                xoffset = double.NaN;
+            }
+        }
+
+        /// <summary>
+        /// Creates a Line conforming to the equation y = mx + b which passes through the points provided
+        /// </summary>
+        /// <param name="points">The points though which the line should pass</param>
+        /// <exception cref="ArgumentException">Thrown when less than 2 points are provided</exception>
+        /// <exception cref="ArgumentException">Thrown when points are not collinear</exception>
+        public StraightLine2D(IEnumerable<Point2D> points) : this(points.ToArray())
+        {
+        }
+
+        /// <summary>
+        /// Creates a Line conforming to the equation y = mx + b which passes through the points provided
+        /// </summary>
+        /// <param name="points">The points though which the line should pass</param>
+        /// <exception cref="ArgumentException">Thrown when less than 2 points are provided</exception>
+        /// <exception cref="ArgumentException">Thrown when points are not collinear</exception>
+        public StraightLine2D(Point2D[] points)
+        {
+            if (points.Length < 2)
+            {
+                throw new ArgumentException("A line must be constructed from at least two points");
+            }
+
+            Gradient = (points[1].Y - points[0].Y) / (points[1].X - points[0].X);
+            if (double.IsInfinity(Gradient))
+            {
+                yoffset = double.NaN; // special case vertical line has no y-intercept
+                xoffset = points[1].X;
+            }
+            else
+            {
+                yoffset = points[1].Y - (Gradient * points[1].X);
+                xoffset = double.NaN;
+            }
+
+            for (int i = 2; i < points.Length; i++)
+            {
+                if (!IsThrough(points[i]))
+                {
+                    throw new ArgumentException("The points provided are not collinear");
+                }
+            }
+        }
+
+        /// <summary>
+        /// Determines if the line is horizontal
+        /// </summary>
+        /// <returns>True if the line is horizontal</returns>
+        public bool IsHorizontal()
+        {
+            return Gradient == 0;
+        }
+
+        /// <summary>
+        /// Determines if the line is vertical
+        /// </summary>
+        /// <returns>returns true if the line is vertical</returns>
+        public bool IsVertical()
+        {
+            return double.IsInfinity(Gradient);
+        }
+
+        /// <summary>
+        /// Tests if the other line is parallel to this one
+        /// </summary>
+        /// <returns>Returns true if lines are parallel</returns>
+        public bool IsParallel(ILine otherLine)
```

Comment:
Making it `public bool IsParalle<T>l(T otherLine) where T : ILine` avoids boxing.

Reply:
ok